### ISSUE TITLE
How to render multiview image parallelly with PyTorch3D?

### ISSUE THREAD

**Issue by FrankZhang0309** [facebookresearch/pytorch3d#231]
We want to change the camera_optimization_with_differentiable_rendering to a multiview version,
that is, the network would optimize the 6D pose of the model instead of the position of the camera.

In this specific case, I have to render `num_of_cameras` image to calculate the total loss. My question is: **If it's possible to render these images parallelly in `forward()` method**, now I did it with a simple `for` loop which makes the network so slow :(

Here is part of my code:
```
    def forward(self):
        self.image = []
        R0, T0 = look_at_view_transform( dist=self.distance, azim=self.azimuth,elev = self.elevation,
                                          up=self.up,at = self.at,device=device)

        R0 = R0.squeeze()
        T0 = T0.transpose(0,1)
        M = torch.cat([R0,T0],dim=1)
        Row4 = torch.Tensor([[0,0,0,1]]).to(device)
        M = torch.cat([M,Row4],dim=0)


        loss = 0
        for i in range(num_cameras):

            #Eg is the initial view matrix of the camera, we multiply it with model matrix
            E = Eg[i].mm(M)
            camera_pose = E.inverse()
            R = camera_pose[:3,:3]
            T = camera_pose[:3,3:]

            # The line we want to do parallelly
            self.image.append(self.renderer(meshes_world=self.meshes.clone(), R=R, T=T))
            image =self.image[i]
            ref =self.image_ref[i]
            loss += torch.sum((image[..., 3] - ref[...,3]) ** 2)

        loss /= num_cameras

        return loss, self.image

```

**Comment by FrankZhang0309** [comment]
Now I tried inputting a batch of `meshes, verts, faces` to  `self.renderer` for rendering multiple images once, but I cannot see a significate speed up comparing with for loop... Are there some suggestions?

Renderer Code:
```
# Initialize an OpenGL perspective camera.
cameras = OpenGLPerspectiveCameras(device=device)

# To blend the 100 faces we set a few parameters which control the opacity and the sharpness of 
# edges. Refer to blending.py for more details. 
blend_params = BlendParams(sigma=1e-4, gamma=1e-4)

raster_settings = RasterizationSettings(
    image_size=256, 
    blur_radius=np.log(1. / 1e-4 - 1.) * blend_params.sigma, 
    faces_per_pixel=20, 
)

# Create a silhouette mesh renderer by composing a rasterizer and a shader. 
silhouette_renderer = MeshRenderer(
    rasterizer=MeshRasterizer(
        cameras=cameras, 
        raster_settings=raster_settings
    ),
    shader=SoftSilhouetteShader(blend_params=blend_params)
)
```

**Comment by nikhilaravi** [comment]
@FrankZhang0309 have a look at this example where we show how to render the same mesh from multiple viewpoints: https://pytorch3d.org/tutorials/render_textured_meshes#6.-Batched-Rendering. 

If you are still experiencing problems please comment again on this issue!

**Comment by FrankZhang0309** [comment]
> @FrankZhang0309 have a look at this example where we show how to render the same mesh from multiple viewpoints: https://pytorch3d.org/tutorials/render_textured_meshes#6.-Batched-Rendering.
> 
> If you are still experiencing problems please comment again on this issue!

Thanks! I read your example and now I can render a batch of images with the same openGLPerspectiveCamera intrinsic(`fov` and `aspect_ratio`).

Then I read pytorch3d source code, I figure out there is no API to render a batch of images with different intrinsics of cameras. Is there any feasible way to do that? 

Or maybe I need to initialize multiple cameras and use a for loop to do that?

**Comment by nikhilaravi** [comment]
Hi @FrankZhang0309 you can have different intrinsics for each mesh in the batch - you can specify a batch of cameras so that the batch size is the same as for the meshes. 
In the textured mesh rendering tutorial linked above we use a batch of cameras - have a look at how this is initialized.  You can specify a list of `fov` and `aspect_ratio` to the relevant camera class.

**Comment by nikhilaravi** [comment]
@FrankZhang0309 if you don't have any further questions please close this issue.

**Comment by FrankZhang0309** [comment]
@nikhilaravi Okay, Thanks for your help! It works. if I have questions, I will reopen it.

**Comment by wang1528186571** [comment]
> Hi @FrankZhang0309 you can have different intrinsics for each mesh in the batch - you can specify a batch of cameras so that the batch size is the same as for the meshes. In the textured mesh rendering tutorial linked above we use a batch of cameras - have a look at how this is initialized. You can specify a list of `fov` and `aspect_ratio` to the relevant camera class.

I tried the batch rendering but it was still very slow.
Also, it may not fully occupy the memory, and it looks like distributed rendering but it is actually streamlined rendering.
Did you solve this problem later?
